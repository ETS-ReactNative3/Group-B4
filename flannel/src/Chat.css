#message_container {
    width: 50%;
    margin: auto;
}

#message {
    margin: auto;
    margin-top: 5px;
    width: 300px;
    border: solid 1px black;
    border-radius: 10px;
    display: flex;
    flex-direction: column;
}

#sender-text {
    margin-top: -5px;
    margin-left: 5px;
    font-size: 18px;
}

#message-text {
    margin-left: 10px;
    margin-bottom: 20px;
}