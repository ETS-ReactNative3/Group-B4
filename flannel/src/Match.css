#current_match {
    width: 96%; 
    height: 50px; 
    background-color: #9ec792;
    border-radius: 5px;
    border: solid 1px #5c6e56;
    display: flex;
    justify-content: center;
    align-items: center;
}

#current_match:hover {
    width: 96%; 
    height: 50px; 
    background-color: #bae9ab;
    border-radius: 5px;
}
