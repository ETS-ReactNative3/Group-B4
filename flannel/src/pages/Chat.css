#message_container {
    width: 50%;
    margin: auto;
}

#chats {

}
#chat_page {
    display: flex; 
    flex-direction: row;
    height: 750px;
    margin-top: 20px;
}

#message {
    margin: auto;
    margin-top: 5px;
    width: 300px;
    border-radius: 10px;
    display: flex;
    flex-direction: column;
    background-color: #f0f0ed;
    border: solid 1px #d9d9d9
}

#sender-text {
    margin-top: 5px;
    margin-left: 5px;
    font-size: 18px;
    width:80%
}

#message-text {
    margin-left: 10px;
    margin-bottom: 20px;
    width:80%
}

body, html {
    background-color: white;
    height: 100%;
  }

#message_window {
    border: solid 1px #d9d9d9;
    width: 50%;
    height: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;
}

#current_chats { 
    width: 150px; 
    height: 100%; 
    border: solid 1px rgb(201, 214, 176);
    background-color: #fefbe9;
    display: flex;
    align-items: center;
    flex-direction: column;
}

#current_chats > * {
    margin: 2px 0;
}


#form_container {
    width: 100%; 
    height: 5%; 
}

#message_list{
    width: 100%;
    height: 95%;
    overflow-x: scroll;
}

#msgform {
    margin: 0 auto;
    width: 50%;
}

#txtbox {
    width: 60%;
    height: 25px;
}

.sendmsg {
    background-color: white; /* Green */
    border: none;
    color: black;
    height: 27px;
    margin-left: 5px;
    border-radius: 3px;
    border: solid 2px hsl(231, 69%, 65%);
    text-align: center;
    text-decoration: none;
    display: inline-block;
    font-size: 16px;
    
}

.sendmsg:hover {
    background-color: hsl(231, 69%, 65%);
    border: solid 1px hsl(231, 69%, 65%);
    color: white
}