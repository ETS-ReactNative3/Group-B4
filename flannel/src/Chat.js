import logo from './logo.svg';
import './App.css';
import './Chat.css';
import './Match.css';
import Match from './Match.js'
import React, { useState, useEffect, useContext} from 'react';
import socketClient  from "socket.io-client";
import {SocketContext, socket} from './socket'

function App({sender, reciever}) {
  const queryParams = new URLSearchParams(window.location.search);
  const from = queryParams.get('from')
  const to = queryParams.get('to')
  const [room, setRoom] = useState(""); //(to + "#" + from).split('#').sort().join("#")
  const [msg, setMsg] = useState(0);
  const [users, setUsers] = useState([{"username": "Henry", "id": "6202ccd95532a766e2477e41"}, {"username": "Brandon", "id": "620306331112997f1a5faffb"}, {"username": "chi", "id": "6204b3f975b3b558c0c773a0"}])
  const [messages, setMessages] = useState([]);//{"timeStamp": 0, msg: "", sender: ""}


  useEffect(() => {
    socket.emit('joinRoom', { username: from, room: room });

    socket.on('messageHistory', (messageHistory) => {
      let temp = [];
      setMessages(() => [])
      messageHistory.forEach((item) => temp.push({"timeStamp": item.timeStamp, "msg": item.msg, "sender": item.sender, "sender_name": item.sender_name}))
      temp.forEach((item) => {
        setMessages((old) => [...old, item])
      })
    })


    socket.on('message', (message) => {
      console.log(message)
      setMessages((old) => [...old, message]);
    })
  }, [socket])


  function sendMsg(e) {
    e.preventDefault();
    let msgObj = {
      room: room, 
      sender: from, 
      data: msg, 
      sender_name: from
    }
    
    let msgString = JSON.stringify(msgObj);
    socket.emit('chat', msgString)
  }
  function handleMsgChange(e) {
    setMsg(e.target.value)
  }

  function updateChat(username, userId) {
    //gives us the username and userId of the selected user we want to chat with 
    socket.emit('leaveRoom', {room: room});
    let tmp_room = (from + "#" + userId).split('#').sort().join("#");
    setRoom(tmp_room)
    socket.emit('joinRoom', { username: from, room: tmp_room});
  }

  return (
    <div id = "chat_page">
      <div id = "current_chats">
        {
          users.map((item, index) => (
            <Match username={item.username} userId={item.id} key={index} callBack={updateChat}></Match>
          ))}
      </div>
      <div className="Message" id = "message_window">
      <div id="message_list">
        {messages.map((message, index) => (
          <div key={index} id = "message_container">
            <div id = "message">
              <div id = "sender-text">
                  <p>{message.sender_name}: </p>
              </div>
              <div id = "message-text">
                {message.msg}
              </div>
            </div>
          </div>
        ))}
      </div>
      <div id = "form_container">
        <form id="msgform">
            <input type = "text" name="message" onChange={handleMsgChange} id="txtbox"></input>
            <button type = "button"  onClick={sendMsg} className = "sendmsg"> Send Msg </button>
        </form>
      </div>  
    </div>
  </div>
    
  );
}

export default App;
