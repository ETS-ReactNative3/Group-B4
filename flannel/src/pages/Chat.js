
import './Chat.css';
import './Match.css';
import Match from './Match.js'
import React, { useState, useEffect, useContext} from 'react';
import socketClient  from "socket.io-client";
import {SocketContext, socket} from '../socket'

function App() {
  const queryParams = new URLSearchParams(window.location.search);
  const from = JSON.parse(localStorage.getItem('user'))._id;
  const userName = JSON.parse(localStorage.getItem('user')).username;
  const [room, setRoom] = useState(""); //(to + "#" + from).split('#').sort().join("#")
  const [msg, setMsg] = useState(0);
  const [users, setUsers] = useState([{"username": "Henry", "id": "62058ecec54374bc9112fc2d"}, {"username": "ishaan", "id": "6205904f308e7d7d8d84075f"}, {"username": "chi", "id": "6204b3f975b3b558c0c773a0"}])
  const [messages, setMessages] = useState([]);//{"timeStamp": 0, msg: "", sender: ""}


  useEffect(async () => {

    const cookies = document.cookie;
    let requestObj = {
      method: 'GET',
      headers: {
          'Content-Type': 'application/json',
          Authorization: cookies,
      },
  };
    //const response = await fetch(`/apihttp://localhost:3000/user?username=aadsfas`, requestObj);
    const response = await fetch(`/api/user/getMatchesList?username=${userName}`, requestObj);

    if(response.status === 401) {
      //render unauth page
    } else {
      const matchList = await response.json();
      setUsers(() => matchList.matches)
    }
    //make an API here to call the list of matched users
    socket.emit('joinRoom', { username: from, room: room });

    socket.on('messageHistory', (messageHistory) => {
      let temp = [];
      setMessages(() => [])
      messageHistory.forEach((item) => temp.push({"timeStamp": item.timeStamp, "msg": item.msg, "sender": item.sender, "sender_name": item.sender_name}))
      temp.forEach((item) => {
        setMessages((old) => [...old, item])
      })
    })


    socket.on('message', (message) => {
      setMessages((old) => [...old, message]);
    })
  }, [socket])


  function sendMsg(e) {
    e.preventDefault();
    let msgObj = {
      room: room, 
      sender: from, 
      data: msg, 
      sender_name: userName
    }
    
    let msgString = JSON.stringify(msgObj);
    socket.emit('chat', msgString)
  }
  function handleMsgChange(e) {
    setMsg(e.target.value)
  }

  function updateChat(username, userId) {
    //gives us the username and userId of the selected user we want to chat with 
    socket.emit('leaveRoom', {room: room});
    let tmp_room = (from + "#" + userId).split('#').sort().join("#");
    setRoom(tmp_room)
    socket.emit('joinRoom', { username: from, room: tmp_room});
  }

  return (
    <div id = "chat_page">
      <div id = "current_chats">
        {
          users.map((item, index) => (
            <Match username={item.username} userId={item.id} key={index} callBack={updateChat}></Match>
          ))}
      </div>
      <div className="Message" id = "message_window">
      <div id="message_list">
        {messages.map((message, index) => (
          <div key={index} id = "message_container">
            <div id = "message">
              <div id = "sender-text">
                  <p>{message.sender_name}: </p>
              </div>
              <div id = "message-text">
                {message.msg}
              </div>
            </div>
          </div>
        ))}
      </div>
      <div id = "form_container">
        <form id="msgform">
            <input type = "text" name="message" onChange={handleMsgChange} id="txtbox"></input>
            <button type = "button"  onClick={sendMsg} className = "sendmsg"> Send Msg </button>
        </form>
      </div>  
    </div>
  </div>
    
  );
}

export default App;
