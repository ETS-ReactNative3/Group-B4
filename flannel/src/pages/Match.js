import React from 'react'
import { callbackify } from 'util';
import './Match.css';
import {Button} from '@mui/material'

function App ({username, userId, callBack}) {
    function chatWithUser(e) {
        e.preventDefault();
        callBack(username, userId);
    }
    return (
        <Button onClick = {chatWithUser} sx= {{padding:1}}>
            {username}
        </Button>
            // <div id = "current_match" onClick={chatWithUser}>
            //     <div id = "user_name">{username}</div> 
            // </div>
    );
}

export default App;