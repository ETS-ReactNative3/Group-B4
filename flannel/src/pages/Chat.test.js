import Chat from './Chat';
import React from 'react';
import { mount, configure, shallow} from 'enzyme';
import ReactDOM from 'react-dom'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Match from './Match.js'
import io from "socket.io-client";
import toJson from 'enzyme-to-json';

const fakeLocalStorage = (function () {
    let store = {};
  
    return {
      getItem: function (key) {
        return store[key] || null;
      },
      setItem: function (key, value) {
        store[key] = value.toString();
      },
      removeItem: function (key) {
        delete store[key];
      },
      clear: function () {
        store = {};
      }
    };
})();

jest.mock("socket.io-client", () => {
    const emit = jest.fn();
    const on = jest.fn();
    const socket = { emit, on };
    return jest.fn(() => socket);
    
  });


describe('chat history', () => {
    beforeAll(() => {
      // socketio.mockClear();
      // socketio().on.mockClear();
      // socketio().emit.mockClear();
        configure({ adapter: new Adapter() });
        Object.defineProperty(window, 'localStorage', {
            value: fakeLocalStorage,
        });
        //now we need to save username and id to local storage 
        window.localStorage.setItem('user', JSON.stringify({
            username: "henry@g.ucla.edu",
            _id: "62058ecec54374bc9112fc2d"
        }))
    })

    it('perform snapshot testing of chat', async ()=> {
        const chat = shallow(<Chat/>);
        expect(chat).toMatchSnapshot();
    })

    it('tests a users match count', async ()=> {
        const chat = shallow(<Chat/>);
        expect(chat.find(Match)).toHaveLength(3);
        expect(chat.find("#chats")).toHaveLength(0);
    })

    it('tests that a user initially has 0 chats', async ()=> {
        const chat = shallow(<Chat/>);
        expect(chat.find("#chats")).toHaveLength(0);
    })

    it('tests that a user can send messages', async ()=> {
        const chat = shallow(<Chat/>);
        chat.find("#name").prop('onKeyDown')({key: 'Enter'})
        expect(chat.find("#chats")).toHaveLength(0);
    })

    it('tests that the message window renders', async () => {
        const chat = shallow(<Chat/>);
        expect(chat.find("#window")).toHaveLength(1);
        expect(chat.find("#current_chats")).toHaveLength(1);
        expect(chat.find("#chat_page")).toHaveLength(1);
    })

    it('tests socket.io', async () => {
      const tmp = io("http://localhost:3000")

      const chat = shallow(<Chat/>);
      expect(tmp.emit).toHaveBeenCalledTimes(0);
      expect(tmp.on).toHaveBeenCalledTimes(0);
        which: 13,
        preventDefault: () => {},
        target: {
          value: ""
        }
      }))
      expect(tmp.emit).toHaveBeenCalledTimes(1);
      expect(tmp.emit).toHaveBeenCalledWith('chat', "{\"room\":\"\",\"sender\":\"62058ecec54374bc9112fc2d\",\"data\":\"\",\"sender_name\":\"henry@g.ucla.edu\"}");
  })
})