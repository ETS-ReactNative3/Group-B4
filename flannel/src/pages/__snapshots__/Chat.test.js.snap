// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat history perform snapshot testing of chat 1`] = `ShallowWrapper {}`;

exports[`chat history tests 1`] = `ShallowWrapper {}`;
